### DISCUSSION TITLE
I'm a moron and can't figure out what to put in the `$redirectUri` field

### DISCUSSION THREAD

**Discussion by Kostas-Mavrogalos** [PHPMailer/PHPMailer#2505] [Q&A]
Ok, so I've been trying to make a contact form, and at the moment I'm at [this](https://github.com/PHPMailer/PHPMailer/wiki/Using-Gmail-with-XOAUTH2#configure-an-oauth2-app) point, at the bottom. I can't figure out what I need to put in the `$redirectUri` field
As far as I've figured, it needs to:
a) have a protocol (https / http)
b) can't be a public IP address
c) needs to points to a script that does what `get_oauth_token.php` does.

But what exactly do I need to put there? `http://localhost/phpmailer/get_oauth_token.php` doesn't work (I hoped it would tbh), and any other variation of the above has the same result.
What worked on you? What am I doing wrong?

**Comment by Synchro** [comment]
`localhost` will only work if your local machine is actually set up to run PHP, and that URL maps to the right script on there. Personally I run nginx as a reverse proxy in front of php-fpm to serve stuff from localhost, however, for the purposes of getting a one-off OAuth key, you can probably use php's built-in server – run `php -S localhost:8000` from within your project folder, check that you can actually run scripts through `http://localhost:8000/` (including the get_oauth one), then paste the path into that google field. When you run the authentication script, google will redirect you to that localhost URL, which will *only* work for you because you're running that endpoint locally, it will then give you the vital keys that you need and with a bit of luck you'll never have to do it again :)

**Reply by Kostas-Mavrogalos** [reply]
Ok, I got about 50% of what you said, let me try to simplify it, so that I'm sure I got what you mean. My machine needs to be able to run PHP to run the script,which makes sense. I should run `php -S localhost:8000` from the project folder of my local machine, right(not the server the site it stored into)??? Cause running that in the Terminal from that folder prints `[Wed Sep 22 10:20:16 2021] PHP 7.4.3 Development Server (http://localhost:8000) started`, and then what do I do? Lastly, what path should I paste into the google field? The one `pwd` prints?
Thanks in advance,and sorry for the dumb questions.

**Reply by Synchro** [reply]
It doesn't matter where the script is run, so long as *your browser* can see it. Google doesn't need to be able to get to your script, only your browser does; anything on `localhost` is not accessible to the outside world, but can be reached from your browser, and google will be quite happy to send you a redirect pointing there. The path in the URL will be relative to your project root, so if your project is in `/Users/myname/Sites/myproject`, you run PHP in there, and by default it will serve whatever is in `/Users/myname/Sites/myproject/index.php` when you request `http://localhost:8000/index.php`, and similarly for any other files in the same folder.

**Reply by Kostas-Mavrogalos** [reply]
Ok, I finally figured out that you meant I would run `php -S localhost:8000` in the project folder in the terminal, then open `http://localhost:8000/index.php` in the browser, my bad for not thinking of that earlier tbh. Now, when I go to `http://localhost:8000/PHPMailer/get_oauth_token.php`, a menu appears with the 3 providers (as the script shows), but when I click on Google, I get redirected to `http://localhost:8000/PHPMailer/get_oauth_token.php?provider=Google` and see an empty page. I have updated the script with the Client ID, Client Secret and Redirect URI as well. Am I doing something wrong?

**Comment by Synchro** [comment]
If there's an error it might be appearing in the PHP log output, or enable error output with `ini_set('display_errors', 1);` in your script.

**Reply by Kostas-Mavrogalos** [reply]
The error is `Warning: require(vendor/autoload.php): failed to open stream: No such file or directory in /home/mark/Documents/Villas_v2/PHPMailer/get_oauth_token.php on line 67`, which would make sense, since I didn't download PHPMailer though Composer, but through the .zip file. Do I need to download Composer and have it in the same directory as the project or is there a workaround?

**Reply by Synchro** [reply]
No, you need to *install* composer on your local machine, and then use it to install PHPMailer in your project. The process of doing that will generate the `vendor` folder and the `autoload.php` file.

Alternatively, alter the file so that it loads the PHPMailer classes manually, as you're probably doing in your other scripts already.

**Reply by Kostas-Mavrogalos** [reply]
The classes that are in `src/PHPMailer.php`, correct?

**Reply by Synchro** [reply]
Yes

**Reply by Kostas-Mavrogalos** [reply]
Another error appears: `Fatal error: Uncaught Error: Class 'League\OAuth2\Client\Provider\Google' not found in /path/to/project/PHPMailer/get_oauth_token.php:104 Stack trace: #0 {main} thrown in /path/to/projectPHPMailer/get_oauth_token.php on line 104`?

**Reply by Synchro** [reply]
Yes, you're missing the packages for OAuth. This is really getting into why not using composer is a bad idea... As soon as you get into anything that's not completely trivial, composer makes it all a lot easier. This is covered in the docs, and the package names it wants are listed in the `suggest` section of `composer.json`. It would be far easier to figure out how to use composer than to figure out how to load these classes manually.

**Reply by Kostas-Mavrogalos** [reply]
So, I downloaded both PHPMailer and the League/OAuth2/... packages, configured everything and now, after i go to `http://localhost:8000/vendor/phpmailer/get_oauth_token.php` and select Google, I select the Gmail account I want, press "Allow" on the next screen and then a PR_END_OF_FILE_ERROR appears on Firefox. Did I do something wrong or is this a Firefox issue?

**Reply by Synchro** [reply]
This might be caused by recent Firefox security changes, in that it may refuse to connect to non-secure endpoints, or this may also be caused by antivirus software, so try disabling that temporarily. Maybe try a different browser too.

**Reply by Kostas-Mavrogalos** [reply]
Opera, Chrome and Tor don't make a difference, sadly. And antivirus was disabled during all that... Any other suggestions? :/

**Reply by Synchro** [reply]
Sorry, no idea.

**Comment by decomplexity** [comment]
As soon as you have ‘allowed’ Google to provide an access token, Google will link  back to your redirectURI (which is the full URI of get_access_token).  If this is not https, Google will probably bounce the connection and leave Firefox hanging. 
Your earlier posts in this thread just showed http.
Note also that the redirectURI you specify to Google when you are setting up OAuth2 must be character-for-character identical with your get_access_token.php's actual URI,
